### DISCUSSION TITLE
What's the difference between partup and WKS/WIC or FIT images?

### DISCUSSION THREAD

**Discussion by mprt** [phytec/partup#84] [Q&A] [answered]
Hi,
can you maybe elaborate a bit on what problem partup is trying to solve?
I'm not sure if I understand when to use partup, since there are also other methods available to package stuff together and define partitions.

What's the position of partup in the tool landscape compared to WKS/WIC (which define a partition layout) and FIT-images (which allow bundling of kernels, DTB and initrd)?

Thanks a lot!
Matthias

**Comment by mschwan-phytec** [comment, marked as answer]
Thanks for asking!

Before we started developing partup, there were some important features, most or even all other partitioning tools were and still are lacking. Most of these cannot be implemented into the existing tools, like WIC or FIT images, as they are out-of-scope (conceptually) of these projects:
* **Dynamic sizes for partitions, depending on the actually used device later on:** Most partitioning tools allow for specifying fixed sizes of their partitions. Sometimes it is also possible to specify a percentage that partition should occupy, but that needs information on the size of the complete disk. However, the size is usually not known during build-time of the image (e.g. when using Yocto) and can be highly variable.
* **Support for eMMC-specific features:** For a lot of our use-cases, we take advantage of the eMMC boot partitions (like `/dev/mmcblk2boot0`) to store the bootloader. WIC and FIT images (and other partitioning tools) do not provide a convenient way to handle these in a complete image. It was always a separate step that needed additional commands to be executed (using `mmc` and alike). There are other eMMC feature we also need, like the eMMC bootbus and hardware reset configuration, some of them being a one-time-programmable bit. Again, WIC (and also FIT) images cannot tackle these features, as this is out-of-scope of their concept.

There were also other motivations, which led us to develop a new tool:
* **Write speed:** Writing WIC images can take quite some time, especially if the root filesystem is larger than a few hundred megabytes. Admittedly, this issue was mostly solved with bmaptools, and writing WIC images with bmaptools is still a little faster than partup (although, not by much). But bmaptools has one major flaw in that it simply *skips* holes (sections with zero-valued bytes). This can be a problem, if there was data on the disk before, because bmaptools may not overwrite these. A common case we encountered was the saved U-Boot environment from a previous system, which may be different/incompatible with the newly flashed system. partup solves this issue by providing an option to clear specific regions on the disk and it also recreates the partition table and formatting completely, overwriting any previous ones.
* **Comfortable image creation and installation:** As mentioned previously with the eMMC specific features, we needed a way to easily create and install an image in a single command each.
* **Utilizing the full space of the disk dynamically:** When having different target devices with different disk sizes, we want to utilize the full space.
* **Small storage size of the image:** WIC images store their content sequentially and raw. In reality this means a WIC image can be several hundred megabytes large (or even gigabytes). Of course, WIC images can be effectively compressed, e.g. with XZ, but compression and decompression can take quite a bit of time. With partup, we use a squashfs, which can be mounted very easily and is fast in reading. The resulting partup package is not as small as a WIC image compressed with XZ, but still significantly smaller than the uncompressed image, of course, and we gain a lot of time as a compromise (because (de)compression of the wic.xz image is slow).
* **Flexibility with different flash types:** WIC images are meant for managed/governed flash devices, like SD cards, eMMCs and SSDs. For raw flash NAND and NOR you have to choose a different image format and flash files individually. Although not implemented yet, partup is designed with a flexible API to allow for flashing raw NAND/NOR devices in the future, if required.

To answer your question when to use partup: You can literally use it always when writing an image to an SD card, eMMC or SSD for embedded devices. ... but also for non-embedded devices if you want, although our main use-case is embedded. :wink:

**Reply by mprt** [reply]
Thank you very much for that exhaustive answer! :-)

**Reply by mschwan-phytec** [reply]
You're welcome! :blush:
